static/js/reactions.js

```diff
@@ -142,26 +142,51 @@ exports.process_reaction_click = function (message_id, local_id) {
     update_ui_and_send_reaction_ajax(message_id, reaction_info);
 };
 
-function full_name(user_id) {
-    if (user_id === page_params.user_id) {
-        return "You (click to remove)";
-    }
-    return people.get_by_user_id(user_id).full_name;
-}
-
 function generate_title(emoji_name, user_ids) {
-    const i = user_ids.indexOf(page_params.user_id);
-    if (i !== -1) {
-        // Move current user's id to start of list
-        user_ids.splice(i, 1);
-        user_ids.unshift(page_params.user_id);
-    }
-    const reacted_with_string = " reacted with :" + emoji_name + ":";
-    const user_names = user_ids.map(full_name);
-    if (user_names.length === 1) {
-        return user_names[0] + reacted_with_string;
+    const usernames = user_ids
+        .filter((user_id) => user_id !== page_params.user_id)
+        .map((user_id) => people.get_by_user_id(user_id).full_name);
+    const current_user_reacted = user_ids.length !== usernames.length;
+
+    const context = {
+        emoji_name: ":" + emoji_name + ":",
+    };
+
+    switch (user_ids.length) {
+        case 1:
+            if (current_user_reacted) {
+                return i18n.t("You (click to remove) reacted with __emoji_name__", context);
+            }
+            context.username = usernames[0];
+            return i18n.t("__username__ reacted with __emoji_name__", context);
+        case 2:
+            if (current_user_reacted) {
+                context.other_username = usernames[0];
+                return i18n.t(
+                    "You (click to remove) and __other_username__ reacted with __emoji_name__",
+                    context,
+                );
+            }
+            context.first_username = usernames[0];
+            context.second_username = usernames[1];
+            return i18n.t(
+                "__first_username__ and __second_username__ reacted with __emoji_name__",
+                context,
+            );
+        default:
+            context.comma_separated_usernames = _.initial(usernames).join(", ");
+            context.last_username = _.last(usernames);
+            if (current_user_reacted) {
+                return i18n.t(
+                    "You (click to remove), __comma_separated_usernames__ and  __last_username__ reacted with __emoji_name__",
```

Comment:
There's a double-space here that we likely don't want, right?

Reply:
Fixed.